src/main/java/com/google/devtools/build/lib/analysis/RunfilesSupport.java

```diff
@@ -495,4 +521,56 @@ public static Path inputManifestPath(Path runfilesDir) {
   public static Path outputManifestPath(Path runfilesDir) {
     return runfilesDir.getRelative(OUTPUT_MANIFEST_BASENAME);
   }
+
+  @Nullable
+  private static Artifact createRepoMappingManifestAction(
+      RuleContext ruleContext, Artifact owningExecutable) {
+    NestedSet<Package> transitivePackages =
+        ruleContext.getTransitivePackagesForRunfileRepoMappingManifest();
+    if (transitivePackages == null) {
+      // For environments where transitive packages are not tracked, we don't have external repos,
+      // so don't build the repo mapping manifest in such cases.
+      return null;
+    }
+    Set<NestedSet.Node> visitedNestedSets = new HashSet<>();
+    Map<RepositoryName, RepositoryMapping> reposAndMappings = new HashMap<>();
+    collectReposAndMappings(transitivePackages, visitedNestedSets, reposAndMappings);
```

Comment:
Honestly not sure; I did it this way since I thought the nested set could be very big and it's not great to hold on to it after analysis, but maybe your concern about performance is more pressing. I'll wait for @lberki to weigh in.

Reply:
Do you have benchmarks for Bazel (since this feature is not enabled in Blaze) that would be able to catch such memory/CPU regressions?

Reply:
+1 for what Fabian says; the nested set is in fact big, but AFAICT we are not disposing of it, since a reference to it just stays in `ConfiguredTargetValue` / `AspectValue` in most of the cases; the notable exception is `--discard_analysis_cache`, but that one is only used within Google where the number of external repositories is zero and where this code is not very useful so meh (holding on to that nested set here would still hurt Google, but see another review comment I'm writing at the moment)

Reply:
(and yes, we do have benchmarks!)

Reply:
Done -- moved to the action itself